### PR TITLE
Adds average argument to AveragePrecision metric

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?

Fixes https://github.com/PyTorchLightning/metrics/issues/471
Adds `average` argument to `AveragePrecision` metric as requested by users. This will change the default output for multiclass and multilabel input from a list with the score per class to instead output the `macro` average.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/functional/classification/average_precision.py [hunk 1]
```diff
@@ -121,14 +158,33 @@ def _average_precision_compute_with_precision_recall(
...
         res.append(-torch.sum((r[1:] - r[:-1]) * p[:-1]))
-    return res
+
+    # Reduce
+    if average in ("macro", "weighted"):
+        res = torch.stack(res)
+        if torch.isnan(res).any():
+            warnings.warn(
+                "Average precision score for one or more classes was `nan`." " Ignoring these classes in average",
+                UserWarning,
```
Comment (Borda) [inline torchmetrics/functional/classification/average_precision.py:168]:
```suggestion
                "Average precision score for one or more classes was `nan`. Ignoring these classes in average",
                UserWarning
```

File: torchmetrics/functional/classification/average_precision.py [hunk 2]
```diff
@@ -121,14 +158,33 @@ def _average_precision_compute_with_precision_recall(
...
+            )
+        if average == "macro":
+            return res[~torch.isnan(res)].mean()
+        else:
+            weights = torch.ones_like(res) if weights is None else weights
+            return (res * weights)[~torch.isnan(res)].sum()
```
Comment (Borda) [inline torchmetrics/functional/classification/average_precision.py:174]:
```suggestion
        weights = torch.ones_like(res) if weights is None else weights
        return (res * weights)[~torch.isnan(res)].sum()
```

File: torchmetrics/functional/classification/average_precision.py [hunk 3]
```diff
@@ -121,14 +158,33 @@ def _average_precision_compute_with_precision_recall(
...
+    elif average is None:
+        return res
+    else:
+        allowed_average = ("micro", "macro", "weighted", None)
+        raise ValueError(f"Expected argument `average` to be one of {allowed_average}" f" but got {average}")
```
Comment (Borda) [inline torchmetrics/functional/classification/average_precision.py:179]:
```suggestion
    allowed_average = ("micro", "macro", "weighted", None)
    raise ValueError(f"Expected argument `average` to be one of {allowed_average}" f" but got {average}")
```

File: torchmetrics/functional/classification/average_precision.py [hunk 4]
```diff
@@ -27,15 +28,38 @@ def _average_precision_update(
...
+    """
+    preds, target, num_classes, pos_label = _precision_recall_curve_update(preds, target, num_classes, pos_label)
+    if average == "micro":
+        if preds.ndim == target.ndim:
+            # Considering each element of the label indicator matrix as a label
+            preds = preds.flatten()
+            target = target.flatten()
+            num_classes = 1
+        else:
+            raise ValueError("Cannot use `micro` average with multi-class input")
```
Comment (Borda) [inline torchmetrics/functional/classification/average_precision.py:52]:
```suggestion
        raise ValueError("Cannot use `micro` average with multi-class input")
```

**Review by Borda — APPROVED** [review]

File: torchmetrics/functional/classification/average_precision.py [hunk 5]
```diff
@@ -27,15 +28,37 @@ def _average_precision_update(
     target: Tensor,
     num_classes: Optional[int] = None,
     pos_label: Optional[int] = None,
+    average: Optional[str] = "macro",
 ) -> Tuple[Tensor, Tensor, int, Optional[int]]:
-    return _precision_recall_curve_update(preds, target, num_classes, pos_label)
+    """Format the predictions and target based on the num_classes, pos_label and average parameter
```
Comment (Borda) [inline torchmetrics/functional/classification/average_precision.py:33]:
```suggestion
    """Format the predictions and target based on the ``num_classes``, ``pos_label`` and ``average`` parameter
```

File: torchmetrics/functional/classification/average_precision.py [hunk 6]
```diff
@@ -67,19 +91,29 @@ def _average_precision_compute(
...
 
     # todo: `sample_weights` is unused
     precision, recall, _ = _precision_recall_curve_compute(preds, target, num_classes, pos_label)
-    return _average_precision_compute_with_precision_recall(precision, recall, num_classes)
+    if average == "weighted":
+        if preds.ndim == target.ndim and target.ndim > 1:
+            weights = target.sum(dim=0).float()
+        else:
+            weights = torch.bincount(target, minlength=num_classes).float()
+        weights = weights / sum(weights)
```
Comment (Borda) [inline torchmetrics/functional/classification/average_precision.py:105]:
```suggestion
        weights = weights / torch.sum(weights)
```

File: torchmetrics/functional/classification/average_precision.py [hunk 7]
```diff
@@ -27,15 +28,37 @@ def _average_precision_update(
...
+        average: reduction method for multi-class or multi-label problems
+    """
+    preds, target, num_classes, pos_label = _precision_recall_curve_update(preds, target, num_classes, pos_label)
+    if average == "micro":
+        if preds.ndim == target.ndim:
+            # Considering each element of the label indicator matrix as a label
+            preds = preds.flatten()
+            target = target.flatten()
+            num_classes = 1
+        raise ValueError("Cannot use `micro` average with multi-class input")
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/average_precision.py:51]:
```suggestion
        else:
            raise ValueError("Cannot use `micro` average with multi-class input")
```

**Review by SeanNaren — APPROVED** [review]
nice!

**Review by tchaton — APPROVED** [review]

**Comment by SeanNaren** [comment]
fine with the wording choice here, but isn't `mean` usually the term? i.e `mAP`?

**Comment by Borda** [comment]
> fine with the wording choice here, but isn't `mean` usually the term? i.e `mAP`?

good point, mAP is mean Average Precision but it is not simple stat mean, but some aggregated mean if I am correct :rabbit: 
@SeanNaren pls have also look at #467

**Comment by SkafteNicki** [comment]
This is the confusing part with the naming of these metrics:
* AveragePrecision (the metric in this PR) refers to the area under the precision-recall curve evaluated user different thresholds. So the name basically means: the average value of precision given all the full range of recall values
* MeanAveragePrecsion or mAP most commonly refers to metric used in object detection. It is also related to the average precision, but the mean is calculated instead over different thresholds of intersection over union between objects (as far as I understand)
So "same same but different".

**Comment by SeanNaren** [comment]
Makes sense, thanks for the clarification guys :)
